contracts/RMRK/extension/erc20Holder/ERC20Holder.sol

```diff
@@ -0,0 +1,127 @@
+// SPDX-License-Identifier: Apache-2.0
+
+pragma solidity ^0.8.18;
+
+import "./IERC20Holder.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+error InvalidValue();
+error InvalidAddress();
+error InsufficientBalance();
+
+abstract contract ERC20Holder is IERC20Holder {
+    mapping(uint256 => mapping(address => uint256)) private balances;
+
+    /**
+     * @inheritdoc IERC20Holder
+     */
+    function balanceOfERC20(
+        address erc20Contract,
+        uint256 tokenId
+    ) external view returns (uint256) {
+        return balances[tokenId][erc20Contract];
+    }
+
+    /**
+     * @inheritdoc IERC20Holder
+     */
+    function transferERC20FromToken(
+        address erc20Contract,
+        uint256 tokenId,
+        address to,
+        uint256 value
+    ) external {
+        if (value == 0) {
+            revert InvalidValue();
+        }
+        if (to == address(0) || erc20Contract == address(0)) {
+            revert InvalidAddress();
+        }
+        if (balances[tokenId][erc20Contract] < value) {
+            revert InsufficientBalance();
+        }
+        _beforeTransferERC20FromToken(erc20Contract, tokenId, to, value);
+        balances[tokenId][erc20Contract] -= value;
+
+        emit TransferredERC20(erc20Contract, tokenId, to, value);
+        _afterTransferERC20FromToken(erc20Contract, tokenId, to, value);
+    }
+
+    /**
+     * @inheritdoc IERC20Holder
+     */
+    function transferERC20ToToken(
+        address erc20Contract,
+        uint256 tokenId,
+        uint256 value
+    ) external {
+        if (value == 0) {
+            revert InvalidValue();
+        }
+        if (erc20Contract == address(0)) {
+            revert InvalidAddress();
+        }
+        _beforeTransferERC20ToToken(erc20Contract, tokenId, msg.sender, value);
+        IERC20(erc20Contract).transferFrom(msg.sender, address(this), value);
```

Comment:
Should we validate the allowance before attempting the transfer? This way we can return our own error, but it also increases contract size..

I think this bypasses the issue of users draining each other 👌

Reply:
About validating allowance, it feels an overkill, it will also spend more gas every time so I'r rather not do it,